### PR TITLE
Fix #1125, use accurate age calculation

### LINKED ISSUE #1125: birthdate date calculation is inaccurate
in models.py in `def age_group_conf(self):`

```
attendee_age = (day - self.birthdate).days / 365.2425
```

we're doing a hokey calculation here and should not be using 365.2425 which probably glosses over some nasty leap-year edge cases.

in subsequent discussions, we should be using something like `dateutils.relativedelta` instead to do this computation.

whoever fixes this should do it Test Driven Development-style:
1) write a unit test showing the current code failing
2) fix that broken test by fixing the code

### PR DESCRIPTION
**Pull request by floraline**
Adds a utility function to calculate a person's age given a birthdate and effective date. This doesn't return a real timespan, so this function shouldn't be used for anything other than finding the age of a person.

I used TDD to implement this function. First, I wrote unit tests to cover a variety of cases that I made up on the spot Then, with the old age code, I ran the tests and found that some of them failed. I've reproduced those results below. Finally, I wrote the new function `get_age_from_birthday` and ran the tests again.

Old code: `(day - self.birthdate).days // 365.2425`

Test failures:
```
get_age_from_birthday(datetime.date(1997, 1, 4), datetime.date(2018, 1, 4))
Expected 21 years, got 20

get_age_from_birthday(datetime.date(2000, 7, 31), datetime.date(2010, 7, 31))
Expected 10 years, got 9
```

That first one could really ruin somebody's day!

New code: `today.year - birthdate.year - ((today.month, today.day) < (birthdate.month, birthdate.day))`

All tests passed successfully.

Please let me know if there are any previously known failure cases for the old code that we can add to the unit tests (or if you just have some ideas!)

Also, I would appreciate if the test cases can be given extra scrutiny. I'm pretty sure they are all correct, but it's late and I'm afraid there might be problems with them that I just am not seeing.

Thanks for reviewing my work. 😄

Fixes #1125

### CODE REVIEW

**Comment by floraline** [comment]
The new code assumes that the age increments by one when the birthdate's month and day have been passed. This means that it acknowledges a February 29 birthday on 2/29 or 3/1, never 2/28.

Is this what we want?

File: uber/tests/models/test_utils.py [hunk 1]
```diff
@@ -93,3 +93,38 @@ def test_charge_log_transaction_no_model(self):
         charge = Charge(amount=1000, description="Test charge")
         Charge.charge_cc(charge, Mock(), 1)
         assert not Charge.stripe_transaction_from_charge.called
+
+
+class TestAgeCalculations:
+
+    @pytest.mark.parametrize('birthdate,today,expected', [
+        # general
```
Comment (kitsuta) [inline uber/tests/models/test_utils.py]: These DOBs and their results will quickly become outdated. You probably want to instead store an intended timedelta (years+months+days) and then pass localized_now() minus the timedelta through to the function for testing.

File: uber/utils.py [hunk 2]
```diff
@@ -548,6 +548,28 @@ def remove_opt(opts, other):
...
+
+    Args:
+        birthdate: A date, should be earlier than the second parameter
+        today:  Optional, age will be found as of this date
+
+    Returns: An integer indicating the age.
+
+    """
+
+    if today == None:
```
Comment (kitsuta) [inline uber/utils.py]: You can skip these two lines by just using the parameter default instead, e.g., `def get_age_from_birthday(birthdate, today=localized_now()):`

File: uber/utils.py [hunk 3]
```diff
@@ -548,6 +548,28 @@ def remove_opt(opts, other):
...
+        today = date.today()
+
+    # Hint: int(True) == 1 and int(False) == 0
```
Comment (kitsuta) [inline uber/utils.py]: I don't understand the purpose of this comment. 😅

File: uber/utils.py (hunk 2 above)
Comment (floraline) [reply uber/utils.py]: I read that this would cause the default value to be evaluated when the function is defined, rather than when it's called. Couldn't that potentially cause undesireable behavior if a system is running for a long period of time? "Now" would gradually become outdated, forever pointing to the day the function was defined. (is this when the server was started?)

Or am I misunderstanding how this works?

File: uber/utils.py (hunk 3 above)
Comment (floraline) [reply uber/utils.py]: Haha, that was supposed to be an explanation for what's going on in the following line. Is it better if I remove it? Or maybe it should just be turned into english? 😆

File: uber/tests/models/test_utils.py (hunk 1 above)
Comment (floraline) [reply uber/tests/models/test_utils.py]: I was actually thinking about this and I was trying to figure out whether this was a problem. I hope you don't mind if I push back a little on this one.

Some of these cases could use a timedelta in that way because they pertain to a specific span of time, and start/end dates aren't important. But most of the test cases pertain to a specific type of date rather than a timespan: they start or end on a leap year, or on Feb 29.

For those test cases, you could only shift the year by increments of four, and that would just be a translation to an equivalent test case. For example, a test case of `2000-1-1` to `2018-1-4` should be equivalent to a test case of `2028-1-1` to `2046-1-4`, or even `2160-1-1` to `2178-1-4`, as long as they all start on a leap year and span the same time frame.

My understanding is that these hard-coded dates should continue to make valid test cases for a developer in the future. I'm open to changing some/all of these test cases per your suggestion, but I don't know right now how I should implement that, or what problem that it would solve.

Comment (kitsuta) [reply uber/tests/models/test_utils.py]: Whoops, no, I just can't read. I didn't notice you were also passing in a date for "today." These are fine!

File: uber/utils.py (hunk 2 above)
Comment (kitsuta) [reply uber/utils.py]: Oh gosh, you're totally right! I think I knew that once and then forgot, my bad!

File: uber/utils.py (hunk 3 above)
Comment (kitsuta) [reply uber/utils.py]: I might remove it, yea -- the line below is totally readable without it, and since it doesn't use integers or booleans that I can see, it's more confusing than not.

To be totally honest... I know that the stack overflow answer this is from said "this works because int(True) is 1 and int(False) is 0" but I have literally no idea why that person said that. 😅

Comment (RobRuana) [reply uber/utils.py]: I agree this particular comment is slightly cryptic, but the code is question is obtuse enough it warrants some kind of explanation. So the initial comment was made with the right spirit! I also often link directly to stack overflow comments instead of trying to explain things in my own words.

The following thoughts are entirely stylistic, entirely personal preference, so I wouldn't ever enforce this in the code base. What you have is already perfectly fine. That being said, let us enter the bikeshed...

In cases like this, I try to make the code itself more self explanatory. What we're really trying to do is determine if the person has had a birthday this year or not. So, I'd try to make the code more explicit in what is happening:

```python
is_today_before_birthday = (today.month, today.day) < (birthdate.month, birthdate.day)
return today.year - birthdate.year - int(is_today_before_birthday)
```

Even more explicit, though perhaps too verbose for an experienced pythonista:

```python
is_today_before_birthday = (today.month, today.day) < (birthdate.month, birthdate.day)
return today.year - birthdate.year - (1 if is_today_before_birthday else 0)
```

You end up with the same number of lines that you had with the comment, but the intent is much clearer. Plus it has the added benefit of keeping the horizontal length of each line shorter – which is nice for those of us who spend all our time on a laptop :smile:

Of course, it's already pretty obvious that `(today.month, today.day) < (birthdate.month, birthdate.day)` is checking if the current day is less than the birthday. I may be overthinking it here 😅

Comment (kitsuta) [reply uber/utils.py]: I actually agree that the two-line version (the first one) is better. I read the original line several times and never quite 'grokked' that it was testing for is_today_before_birthday. Of course, maybe I just was too tired at the time...
